src/dscheck/multicore_magic.ml

```diff
@@ -0,0 +1,44 @@
+module Atomic = Dscheck.TracedAtomic
+
+let copy_as_padded = Fun.id
+let make_padded_array = Array.make
+let length_of_padded_array = Array.length
+let length_of_padded_array_minus_1 xs = Array.length xs - 1
+
+module Transparent_atomic = struct
+  include Atomic
+
+  let make_contended = make
+  let fenceless_get = get
+  let fenceless_set = set
+end
+
+let[@inline] fenceless_get (atomic : 'a Atomic.t) =
+  !(Sys.opaque_identity (Obj.magic atomic : 'a ref))
+
+let[@inline] fenceless_set (atomic : 'a Atomic.t) value =
+  (Obj.magic atomic : 'a ref) := value
+
+let[@inline] fence atomic = Atomic.fetch_and_add atomic 0 |> ignore
+
+module Atomic_array = struct
+  type 'a t = 'a Atomic.t array
+
+  let[@inline] at (type a) (xs : a t) i : a Atomic.t =
+    (* ['a t] does not contain [float]s. *)
+    Obj.magic (Array.unsafe_get (Obj.magic xs : a ref array) i)
```

Comment:
This will also cause problems.  We should change to:
```ocaml
Atomic.get xs.(i)
```

Reply:
I guess you meant `Array.get xs i` .